src/display/canvas.js

```diff
@@ -3104,46 +3069,140 @@ class CanvasGraphics {
+    // lineWidth in order to have both thicknesses greater or equal
+    // to 1 after transform.
+    if (!this._cachedScaleForStroking) {
+      const { lineWidth } = this.current;
+      const m = this.ctx.mozCurrentTransform;
+      let scaleX, scaleY;
+
+      // We must take into account that outputScale is a part of the
+      // current transform.
+      if (m[1] === 0 && m[2] === 0) {
+        // Fast path
+        const normX = Math.abs(m[0]);
+        const normY = Math.abs(m[3]);
+        if (lineWidth === 0) {
+          scaleX = 1 / normX;
+          scaleY = 1 / normY;
+        } else {
+          const scaledXLineWidth = normX * lineWidth;
+          const scaledYLineWidth = normY * lineWidth;
+          scaleX =
+            scaledXLineWidth < this.outputScaleX
+              ? this.outputScaleX / scaledXLineWidth
+              : 1;
+          scaleY =
+            scaledYLineWidth < this.outputScaleY
+              ? this.outputScaleY / scaledYLineWidth
+              : 1;
+        }
       } else {
-        // Matrix is non-invertible.
-        this._cachedGetSinglePixelWidth = 1;
+        // A pixel (base (x, y)) is transformed by M into a parallelogram:
+        //  - its area is |det(M)|;
+        //  - heightY (orthogonal to Mx) has a length: |det(M)| / norm(Mx);
+        //  - heightX (orthogonal to My) has a length: |det(M)| / norm(My).
+        // heightX and heightY are the thicknesses of the transformed pixel
+        // and they must be both greater or equal to 1.
+        const absDet = Math.abs(m[0] * m[3] - m[2] * m[1]);
+        const normX = Math.hypot(m[0], m[1]);
+        const normY = Math.hypot(m[2], m[3]);
+        if (lineWidth === 0) {
+          scaleX = normY / absDet;
+          scaleY = normX / absDet;
+        } else {
+          const baseArea = lineWidth * absDet;
+          const scaledNormY = normY * this.outputScaleX;
+          const scaledNormX = normX * this.outputScaleY;
+          scaleX = scaledNormY > baseArea ? scaledNormY / baseArea : 1;
+          scaleY = scaledNormX > baseArea ? scaledNormX / baseArea : 1;
+        }
       }
+      this._cachedScaleForStroking = [scaleX, scaleY];
     }
+    return this._cachedScaleForStroking;
+  }
 
-    return this._cachedGetSinglePixelWidth;
+  // Rescale before stroking in order to have a final lineWidth
+  // with both thicknesses greater or equal to 1.
+  rescaleAndStroke() {
+    const { ctx } = this;
+    const { lineWidth } = this.current;
+    const [scaleX, scaleY] = this.getScaleForStroking();
+    if (scaleX === 1 && scaleY === 1) {
+      ctx.lineWidth = lineWidth || 1;
+      ctx.stroke();
+      return;
+    }
+
+    const savedMatrix = ctx.mozCurrentTransform.slice();
+    const savedDashes = ctx.getLineDash().slice();
+    const savedDashOffset = ctx.lineDashOffset;
+
+    ctx.scale(scaleX, scaleY);
+    ctx.lineWidth = lineWidth || 1;
+
+    // How the dashed line is rendered depends on the current transform...
```

Comment:
Can we make this chunk instead call rescaleAndStrokeNoSave or add another function that they both call. Having two code paths to keep in sync can be problematic.